Gpu Time Correlation

GrantComm opened the pull request:
Highlights corresponding VkBeginCommandBuffer, Frame, or VkBeginRenderPass node when item is selected in the gpu time tab.

https://github.com/user-attachments/assets/d849048d-1205-4559-96ee-3de41bcdff78

GrantComm commented:
> ` [4f6b8ef](/google/dive/pull/396/commits/4f6b8ef5ab3fbf2658be8d4a80a19912ac3dba61)`

Sorry yes, please start at [4f6b8ef](https://github.com/google/dive/pull/396/commits/4f6b8ef5ab3fbf2658be8d4a80a19912ac3dba61). Updated the description. This change is based on #389

GrantComm commented:
> I'm a bit confused by the PR description that says to start reviewing from [92e7b2f](https://github.com/google/dive/commit/92e7b2f39d0764e6db99b007302580ed3ff6ffa9), it seems like there are important changes from before that point? Is this PR changes based off another pending PR?
> 
> EDIT: I checked #389 and it seems like there are two additional commits in this change, so should I be reviewing beginning from [4f6b8ef](https://github.com/google/dive/commit/4f6b8ef5ab3fbf2658be8d4a80a19912ac3dba61)?

Sorry yes, please start at [4f6b8ef](https://github.com/google/dive/pull/396/commits/4f6b8ef5ab3fbf2658be8d4a80a19912ac3dba61). Updated the description. This change is based on https://github.com/google/dive/pull/389

GrantComm commented on ui/gpu_timing_tab_view.cpp:
```diff
@@ -153,4 +158,35 @@ void GpuTimingTabView::OnEventSelectionChanged(const QModelIndex &model_index)
+        QItemSelectionModel *selection_model = m_table_view->selectionModel();
+        if (selection_model)
+        {
+            selection_model->clear();
```
It essentially the same. ClearSelection() just explicitly gets the selection model for the table and clears it. I added ClearSelection() to OnEventSelectionChanged() and removed the m_ignore_selection_changes. Those were needed to prevent infinite looping between signals but I refactored the selection process in the main window so they are unneeded now.

angela28chen replied:
That's probably the better approach (fixing the infinite signal loop), thanks!

angela28chen commented:
I spent some time thinking about your proposed setup with `m_gpu_time_correlation_from_command_hierarchy` to explicitly block signals and solve the issue of having infinite signal loop triggered by the selection changing for these related tabs (vulkan command hierarchy, pm4 tree, gpu timing). I think rather than use this approach, we should consider the fact that the `currentChanged` signal should only be produced when the selected element has changed. To give an example:

1. User clicks on GPU time row and `GpuTimingTabView::OnSelectionChanged` is triggered
2. Emits `GpuTimingTabView::GpuTimingDataSelected` for a valid row `r`
3. `MainWindow::OnGpuTimingDataSelected` is triggered
4. Selection is cleared for other views (vulkan and pm4)
5. If a correlated row is found, then main window `setCurrentIndex()` is called
6. Eventually comes back around to triggering GpuTimingTabView::OnEventSelectionChanged

I think the issue here is step 4. If we are more sparing about clearing selection, then the signal 5 may not trigger a `currentChanged` if the selection is the same as before. My guess is that if we set it up so that the selection is cleared only when a correlated row is not found, this would terminate the loop once the selections (valid row or none) in the 3 tabs are consistent with each other. I think this should apply to the relationship between the vulkan and pm4 tabs too. 

Please let me know what you think.

https://doc.qt.io/qt-6/qitemselectionmodel.html#currentChanged

GrantComm commented:
> I spent some time thinking about your proposed setup with `m_gpu_time_correlation_from_command_hierarchy` to explicitly block signals and solve the issue of having infinite signal loop triggered by the selection changing for these related tabs (vulkan command hierarchy, pm4 tree, gpu timing). I think rather than use this approach, we should consider the fact that the `currentChanged` signal should only be produced when the selected element has changed. To give an example:
> 
> 1. User clicks on GPU time row and `GpuTimingTabView::OnSelectionChanged` is triggered
> 2. Emits `GpuTimingTabView::GpuTimingDataSelected` for a valid row
> 3. `MainWindow::OnGpuTimingDataSelected` is triggered
> 4. Selection is cleared for other views (vulkan and pm4)
> 5. If a vulkan row that correlates with the gpu row is found, then the vulkan table `setCurrentIndex()` is called
> 6. Eventually that comes back around to triggering GpuTimingTabView::OnEventSelectionChanged, looping back to step 1
> 
> I think the issue here is step 4. If we are more sparing about clearing selection, then step 5 may not trigger a `currentChanged` if the selection is the same as before. My guess is that if we set it up so that the selection is cleared ONLY when a correlated row is not found, this would terminate the loop once the selections (valid row or none) across the 3 tabs are consistent with each other. I think this should apply to the relationship between the vulkan and pm4 tabs too.
> 
> Please let me know what you think.
> 
> https://doc.qt.io/qt-6/qitemselectionmodel.html#currentChanged

I have updated the correlation and removed m_gpu_time_correlation_from_command_hierarchy. I still think the signals should be blocked due in the cases where a gpu time node is selected after draw correlation has occurred though. In that case the views would be cleared and if there is a corresponding begincmdbuffer/beginrenderpass node it would be selected, if not just the selected node would be highlighted.

I changed the use of booleans to QSignalBlocker on the selection models. It blocks the signal on the selection model until it falls out of scope. https://doc.qt.io/qt-6/qsignalblocker.html
